wb/test/sim_main.cpp

```diff
@@ -0,0 +1,211 @@
+                    + ((data & 0xFF) << 48)
+                    + ((data & 0xFF) << 56);
+        case 1: return (data & 0xFFFF)
+                    + ((data & 0xFFFF) << 16)
+                    + ((data & 0xFFFF) << 32)
+                    + ((data & 0xFFFF) << 48);
+        case 2: return (data & 0xFFFFFFFF)
+                    + ((data & 0xFFFFFFFF) << 32);
+        default: return data;
+    }
+}
+
+// After calling timer_tick, the sim time will stop right before the next rising clk edge.
+// This way users can peek the final results of a cycle (like the SystemVerilog
+// Postponed Region).
+void timer_tick(Vtop *dut, VerilatedContext *contextp, VerilatedFstC *tfp) {
+    dut->eval();
+    tfp->dump(contextp->time());
+    contextp->timeInc(1);
+
+    dut->clk_i = 0;
+    dut->eval();
+    tfp->dump(contextp->time());
+    contextp->timeInc(1);
+}
+// After users finish reading the final results of a cycle, they call timer_eval to move to
+// the next time slot.
+// After calling time_eval, the signals in DUT will be computed. Users can then specify the
+// new input signals to the DUT. This resembles the SystemVerilog Reactive Region.
+void timer_eval(Vtop *dut) {
+    dut->clk_i = 1;
+    dut->eval();
+}
+
+
+int main(int argc, char* argv[]) {
+    std::unique_ptr<VerilatedContext> contextp(new VerilatedContext);
+    contextp->commandArgs(argc, argv);
+    std::unique_ptr<VerilatedFstC> tfp(new VerilatedFstC);
+    contextp->traceEverOn(VM_TRACE_FST);
+    std::unique_ptr<Vtop> dut(new Vtop(contextp.get()));
+
+    dut->trace(tfp.get(), 10);
+    Verilated::mkdir("waveforms");
+    tfp->open("waveforms/wave.fst");
+    contextp->time(0);
+    
+    timer_eval(dut.get());
+    dut->reset_i = 1;
+    dut->c_lce_id_i = 0;
+    dut->c_did_i = 0;
+    timer_tick(dut.get(), contextp.get(), tfp.get());
+
+    timer_eval(dut.get());
+    dut->reset_i = 0;
+    timer_tick(dut.get(), contextp.get(), tfp.get());
+    
+    bool error = false;
+    for (int i = 0; i < TEST_SIZE; i++) {
+        // generate a command
+        VL_SIG8(size, 2, 0) = dice() & 0x3;
+        VL_SIG64(addr, 39, 0) = dice() & 0xFFFFFFFFF8;
+        VL_SIG8(msg_type, 3, 0) = gen_msg_type(addr);
+        VL_SIG64(data_cmd, 63, 0) = replicate(dice(), size);
+        BP_cmd cmd_in(size, addr, msg_type, data_cmd);
+
+        // input the command to the master adapter
+        timer_eval(dut.get());
+        dut->m_mem_cmd_header_i = cmd_in.header;
+        dut->m_mem_cmd_data_i = cmd_in.data;
+        dut->m_mem_cmd_v_i = 1;
+        dut->c_mem_cmd_ready_i = 0;
+        timer_tick(dut.get(), contextp.get(), tfp.get());
+        timer_eval(dut.get());
+        dut->m_mem_cmd_v_i = 0;
+
+        // wait a few cycles to check backpressure
+        int delay = dice() % 8;
+        for (int j = 0; j < delay; j++) {
+            timer_tick(dut.get(), contextp.get(), tfp.get());
+            timer_eval(dut.get());
+        }
+        dut->c_mem_cmd_ready_i = 1;
```

Comment:
Seems like this is missing backpressure cases such as back to back writes or reads without acknowledgements